modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/js/editor/BaseEditor.js

```diff
@@ -62,7 +63,7 @@ const BaseEditor = forwardRef(
 		}, []);
 
 		useEffect(() => {
-			if (!initialConfig.editorTransformerURLs) {
+			if (deployed || !initialConfig.editorTransformerURLs) {
```

Comment:
I think the key to better understand this issue is in [this comment](https://liferay.atlassian.net/browse/PTR-5174?focusedCommentId=2695472) ([PTR-5174](https://liferay.atlassian.net/browse/PTR-5174)) which says:

> _Therefore, as this is also a form field, the actions executed in it are traced through usePropagateAction , every time something like ‘field_focus’(click case) or ‘field_blur' is executed, it’s sending an ‘new’ editorConfig object all the way to the field component(RichText), which will eventually cause the triggering of the mentioned useEffect in BasicEditor, even if the values within the initialConfig object it remain unchanged_

Reply:
I was also thinking on making the effect run just once, on first render. However, this will prevent CX from running on subsequent renders, which can be unexpected if CX has some logic that can produce different results over time (imagine a CX that switches to "night" editor theme after 21hrs, or just fecthes per-user configuration from a separate repository)

We could solve this by delegating the responsibility of being idempotent to each and every CX instance. This means that, no matter how many times it's run, the end result would be the same (e.g. don't add the button if it's already added in the incoming config). To me, putting this burden on CX implementor is not reasonable and would instead denote some odd state management on editor side.

The other option is to hide the config transformations into a piece of state, controlled only from within BaseEditor. Whereas `initialConfig` is given from outside, BaseEditor would do as follows:

- Deep clone initialConfig into a new `internalConfig` object
- Apply CX on `internalConfig`
- Instantiate editor and pass `internalConfig` to it

This way, `config` (or `initialConfig`) would be controlled but the part concerning CX would be uncontrolled

Does that make sense?

Reply:
To me it seems like the best approach is changing this to the code snippet in @markocikos's comment https://github.com/liferay-frontend/liferay-portal/pull/4430#discussion_r1768771889 since I'd think loading the CX should only need to happen once.

I tested doing a deep cloning `initialConfig` into a new `internalConfig` object (maybe not exactly what you were thinking) which does result in a consistent `transformedConfig`, but if I still leave `[initialConfig]` as a dependency of the useEffect, the editor will still re-render on every interaction:

```
const internalConfig = JSON.parse(JSON.stringify(initialConfig));

useEffect(() => {
	if (!initialConfig.editorTransformerURLs) {
		return;
	}

	setLoading(true);

	loadEditorClientExtensions({
		config: internalConfig,
		onLoad: ({transformedConfig}) => {
			setConfig(transformedConfig);

			setLoading(false);
		},
	});

	// eslint-disable-next-line react-hooks/exhaustive-deps
}, [initialConfig]);
```

But if this passed in `config` in `loadEditorClientExtensions` is mutating the original object, I feel the fix should be somewhere in there, which we can do in a separate ticket if it becomes an issue.

As a first step though, it seems right having `loadEditorClientExtensions` only called once.